torch/csrc/jit/script/tree_views.h

```diff
@@ -20,423 +91,422 @@ struct TreeView {
     return tree_;
   }
-
- protected:
-  TreeRef tree_;
-};
-
-template <typename T>
-struct ListViewIterator {
-  ListViewIterator(TreeList::const_iterator it) : it(it) {}
-  bool operator!=(const ListViewIterator& rhs) const {
-    return it != rhs.it;
-  }
-  T operator*() const {
-    return T(*it);
-  }
-  void operator++() {
-    ++it;
+  const TreeRef& get() const {
+    return tree_;
   }
-  void operator--() {
-    --it;
+  int kind() const {
+    return tree_->kind();
   }
 
- private:
-  TreeList::const_iterator it;
+protected:
+  const TreeRef& subtree(std::size_t i) const {
+    return tree_->trees().at(i);
+  }
+  TreeRef tree_;
 };
 
 template <typename T>
-struct ListView : public TreeView {
-  ListView(const TreeRef& tree) : TreeView(tree) {
+struct List : public TreeView {
+  struct Iterator {
+    Iterator(TreeList::const_iterator it) : it(it) {}
+    bool operator!=(const Iterator& rhs) const { return it != rhs.it; }
+    T operator*() const { return T(*it); }
+    void operator++() { ++it; }
+    void operator--() { --it; }
+
+  private:
+    TreeList::const_iterator it;
+  };
+  typedef Iterator iterator;
+  typedef Iterator const_iterator;
+
+  List(const TreeRef& tree) : TreeView(tree) {
     tree->match(TK_LIST);
   }
-  typedef ListViewIterator<T> iterator;
-  typedef ListViewIterator<T> const_iterator;
   iterator begin() const {
     return iterator(tree_->trees().begin());
   }
   iterator end() const {
     return iterator(tree_->trees().end());
   }
   T operator[](size_t i) const {
-    return T(tree_->trees().at(i));
+    return T(subtree(i));
   }
   TreeRef map(std::function<TreeRef(const T&)> fn) {
     return tree_->map([&](TreeRef v) { return fn(T(v)); });
   }
+  static List create(const SourceRange& range, const std::vector<T>& subtrees) {
+    TreeList type_erased_sub {subtrees.begin(), subtrees.end()};
+    return List(Compound::create(TK_LIST, range, std::move(type_erased_sub)));
+  }
   size_t size() const {
     return tree_->trees().size();
   }
 };
 
 template <typename T>
-struct OptionView : public TreeView {
-  explicit OptionView(const TreeRef& tree) : TreeView(tree) {
+struct Maybe : public TreeView {
+  explicit Maybe(const TreeRef& tree) : TreeView(tree) {
     JIT_SCRIPT_ASSERT(tree, tree->kind() == TK_OPTION);
   }
+  /* implicit */ Maybe(const T& tree) : TreeView(tree) {}
   bool present() const {
     return tree_->trees().size() > 0;
   }
   T get() const {
-    JIT_SCRIPT_ASSERT(tree_, present());
-    return T(tree_->trees()[0]);
+    return T(tree_->trees().at(0));
   }
   TreeRef map(std::function<TreeRef(const T&)> fn) {
     return tree_->map([&](TreeRef v) { return fn(T(v)); });
   }
 };
 
 struct Ident : public TreeView {
-  // each subclass of TreeView provides:
-  // 1. a constructor that takes a TreeRef, and matches it to the right type.
   explicit Ident(const TreeRef& tree) : TreeView(tree) {
-    tree_->match(TK_IDENT, name_);
+    tree_->match(TK_IDENT);
```

Comment:
These match statements should check that number of subtrees is correct. Otherwise errors in create methods won't be caught until the first use, which will be very far away.